lib/jwt/jwt.go

```diff
@@ -0,0 +1,284 @@
+func New(config *Config) (*Key, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	privateKey, err := parsePrivateKey(config.PrivateKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	publicKey, ok := privateKey.Public().(*rsa.PublicKey)
+	if !ok {
+		return nil, trace.BadParameter("invalid public key")
+	}
+
+	return &Key{
+		Config:     config,
+		publicKey:  publicKey,
+		privateKey: privateKey,
+	}, nil
+}
+
+// NewPublic is a JWT key that can only be used to verify tokens.
+func NewPublic(config *Config) (*Key, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	publicKey, err := parsePublicKey(config.PublicKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &Key{
+		Config:    config,
+		publicKey: publicKey,
+	}, nil
+}
+
+// SignParams are the claims to be embedded within the JWT token.
+type SignParams struct {
+	// Username is the Teleport identity.
+	Username string
+
+	// Roles are the roles assigned to the user within Teleport.
+	Roles []string
+
+	// Expiry is time to live for the token.
+	Expiry time.Duration
+}
+
+// Check verifies all the values are valid.
+func (p *SignParams) Check() error {
+	if p.Username == "" {
+		return trace.BadParameter("missing username")
+	}
+	if len(p.Roles) == 0 {
+		return trace.BadParameter("missing roles")
+	}
+	if p.Expiry == 0 {
+		return trace.BadParameter("expiry required")
+	}
+
+	return nil
+}
+
+// Sign will return a signed JSON token with the passed in claims embedded
+// within the token.
+func (k *Key) Sign(p *SignParams) (string, error) {
+	if k.privateKey == nil {
+		return "", trace.BadParameter("can not sign token with non-signing key")
+	}
+	if err := p.Check(); err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	signingKey := jose.SigningKey{
+		Algorithm: jose.RS256,
+		Key:       k.privateKey,
+	}
+
+	sig, err := jose.NewSigner(signingKey, (&jose.SignerOptions{}).WithType("JWT"))
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	claims := jwtClaims{
+		Claims: josejwt.Claims{
+			Subject:   p.Username,
+			Issuer:    "Gravitational Teleport",
+			NotBefore: jwt.NewNumericDate(k.Clock.Now().Add(-10 * time.Second)),
+			Expiry:    jwt.NewNumericDate(k.Clock.Now().Add(p.Expiry)),
+		},
+		Username: p.Username,
+		Roles:    p.Roles,
+	}
+	token, err := josejwt.Signed(sig).Claims(claims).CompactSerialize()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	return token, nil
+}
+
+// Verify will validate the passed in JWT token.
+func (k *Key) Verify(raw string) (*jwtClaims, error) {
+	if k.publicKey == nil {
+		return nil, trace.BadParameter("can not verify token without public key")
+	}
+
+	tok, err := josejwt.ParseSigned(raw)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var out jwtClaims
+	if err := tok.Claims(k.publicKey, &out); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	expectedClaims := jwt.Expected{
```

Comment:
also verify `Issuer`, `Signer` and `Audience`

Reply:
Signer will be verified in the `out.Validate()`?

Reply:
sorry, meant `Subject`, not `Signer`